// bcxjs-cores/dist/serializer/src/operations.js
"use strict";

exports.__esModule = true;
exports.stealth_memo_data = exports.signed_transaction = exports.transaction = exports.asset_claim_fees = exports.asset_settle_cancel = exports.transfer_from_blind = exports.blind_transfer = exports.blind_input = exports.transfer_to_blind = exports.blind_output = exports.stealth_confirmation = exports.override_transfer = exports.balance_claim = exports.assert = exports.block_id_predicate = exports.asset_symbol_eq_lit_predicate = exports.account_name_eq_lit_predicate = exports.custom = exports.worker_create = undefined;
exports.burn_worker_initializer = exports.vesting_balance_worker_initializer = exports.refund_worker_initializer = exports.vesting_balance_withdraw = exports.vesting_balance_create = exports.cdd_vesting_policy_initializer = exports.linear_vesting_policy_initializer = exports.committee_member_update_global_parameters = exports.chain_parameters = exports.committee_member_update = exports.committee_member_create = exports.withdraw_permission_delete = exports.withdraw_permission_claim = exports.withdraw_permission_update = exports.withdraw_permission_create = exports.proposal_delete = exports.proposal_update = exports.proposal_create = exports.op_wrapper = exports.witness_update = exports.witness_create = exports.asset_publish_feed = exports.price_feed = exports.asset_global_settle = exports.asset_settle = exports.asset_fund_fee_pool = exports.asset_reserve = exports.asset_issue = exports.asset_update_feed_producers=exports.asset_update_restricted = exports.asset_update_bitasset = exports.asset_update = exports.asset_create = exports.bitasset_options = exports.asset_options = exports.price = exports.account_transfer = exports.account_upgrade = exports.account_update =exports.contract_create=exports.call_contract_function=exports.register_nh_asset_creator=exports.create_world_view=exports.relate_world_view=exports.create_nh_asset=exports.update_collateral_for_gas=exports.delete_nh_asset=exports.transfer_nh_asset=exports.create_nh_asset_order=exports.cancel_nh_asset_order=exports.fill_nh_asset_order=exports.revise_contract=exports.crontab_create=exports.crontab_cancel=exports.crontab_recover=exports.account_create = exports.account_options = exports.authority = exports.fill_order = exports.call_order_update = exports.limit_order_cancel = exports.limit_order_create = exports.transfer = exports.memo_data = exports.signed_block_header = exports.block_header = exports.signed_block = exports.processed_transaction = exports.asset = exports.void_result = exports.fee_schedule = exports.asset_claim_fees_operation_fee_parameters = exports.asset_settle_cancel_operation_fee_parameters = exports.transfer_from_blind_operation_fee_parameters = exports.blind_transfer_operation_fee_parameters = exports.transfer_to_blind_operation_fee_parameters = exports.override_transfer_operation_fee_parameters = exports.balance_claim_operation_fee_parameters = exports.assert_operation_fee_parameters = exports.custom_operation_fee_parameters = exports.worker_create_operation_fee_parameters = exports.vesting_balance_withdraw_operation_fee_parameters = exports.vesting_balance_create_operation_fee_parameters = exports.committee_member_update_global_parameters_operation_fee_parameters = exports.committee_member_update_operation_fee_parameters = exports.committee_member_create_operation_fee_parameters = exports.withdraw_permission_delete_operation_fee_parameters = exports.withdraw_permission_claim_operation_fee_parameters = exports.withdraw_permission_update_operation_fee_parameters = exports.withdraw_permission_create_operation_fee_parameters = exports.proposal_delete_operation_fee_parameters = exports.proposal_update_operation_fee_parameters = exports.proposal_create_operation_fee_parameters = exports.witness_update_operation_fee_parameters = exports.witness_create_operation_fee_parameters = exports.asset_publish_feed_operation_fee_parameters = exports.asset_global_settle_operation_fee_parameters = exports.asset_settle_operation_fee_parameters = exports.asset_fund_fee_pool_operation_fee_parameters = exports.asset_reserve_operation_fee_parameters = exports.asset_issue_operation_fee_parameters = exports.asset_update_feed_producers_operation_fee_parameters = exports.asset_update_bitasset_operation_fee_parameters = exports.asset_update_operation_fee_parameters = exports.asset_create_operation_fee_parameters = exports.account_transfer_operation_fee_parameters = exports.account_upgrade_operation_fee_parameters = exports.account_update_operation_fee_parameters = exports.account_create_operation_fee_parameters = exports.fill_order_operation_fee_parameters = exports.call_order_update_operation_fee_parameters = exports.limit_order_cancel_operation_fee_parameters = exports.limit_order_create_operation_fee_parameters = exports.transfer_operation_fee_parameters = exports.operation = undefined;

var _types = require("./types");

var _types2 = _interopRequireDefault(_types);

var _serializer = require("./serializer");

var _serializer2 = _interopRequireDefault(_serializer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var uint8 = _types2.default.uint8,
    uint16 = _types2.default.uint16,
    uint32 = _types2.default.uint32,
    int64 = _types2.default.int64,
    int32 = _types2.default.int32,
    double=_types2.default.double,
    varint32 = _types2.default.varint32,
    uint64 = _types2.default.uint64,
    string = _types2.default.string,
    bytes = _types2.default.bytes,
    bool = _types2.default.bool,
    array = _types2.default.array,
    fixed_array = _types2.default.fixed_array,
    protocol_id_type = _types2.default.protocol_id_type,
    nh_protocol_id_type = _types2.default.nh_protocol_id_type,
    object_id_type = _types2.default.object_id_type,
    vote_id = _types2.default.vote_id,
    future_extensions = _types2.default.future_extensions,
    static_variant = _types2.default.static_variant,
    lua_types = _types2.default.lua_types,
    table=_types2.default.table,

    map = _types2.default.map,
    pair = _types2.default.pair,

    
    set = _types2.default.set,
    public_key = _types2.default.public_key,
    address = _types2.default.address,
    time_point_sec = _types2.default.time_point_sec,
    optional = _types2.default.optional;


future_extensions =string; //_types2.default.void;

/*
When updating generated code
Replace:  operation = static_variant [
with:     operation.st_operations = [

Delete:
public_key = new Serializer(
    "public_key"
    key_data: bytes 33
)

*/
// Place-holder, their are dependencies on "operation" .. The final list of
// operations is not avialble until the very end of the generated code.
// See: operation.st_operations = ...
var operation = static_variant();
// module.exports["operation"] = operation;

exports.operation = operation;
// For module.exports

var Serializer = function Serializer(operation_name, serilization_types_object) {
    return new _serializer2.default(operation_name, serilization_types_object);
    // return module.exports[operation_name] = s;
};

// Custom-types follow Generated code:

// ##  Generated code follows
// # programs/js_operation_serializer > npm i -g decaffeinate
// ## -------------------------------
var transfer_operation_fee_parameters = exports.transfer_operation_fee_parameters = new Serializer("transfer_operation_fee_parameters", { fee: uint64,
    price_per_kbyte: uint32 });

var limit_order_create_operation_fee_parameters = exports.limit_order_create_operation_fee_parameters = new Serializer("limit_order_create_operation_fee_parameters", { fee: uint64 });

var limit_order_cancel_operation_fee_parameters = exports.limit_order_cancel_operation_fee_parameters = new Serializer("limit_order_cancel_operation_fee_parameters", { fee: uint64 });

var call_order_update_operation_fee_parameters = exports.call_order_update_operation_fee_parameters = new Serializer("call_order_update_operation_fee_parameters", { fee: uint64 });

var fill_order_operation_fee_parameters = exports.fill_order_operation_fee_parameters = new Serializer("fill_order_operation_fee_parameters");

var account_create_operation_fee_parameters = exports.account_create_operation_fee_parameters = new Serializer("account_create_operation_fee_parameters", { basic_fee: uint64,
    premium_fee: uint64,
    price_per_kbyte: uint32 });

var account_update_operation_fee_parameters = exports.account_update_operation_fee_parameters = new Serializer("account_update_operation_fee_parameters", { fee: int64,
    price_per_kbyte: uint32 });

//43
var contract_create_operation_fee_parameters = exports.contract_create_operation_fee_parameters = new Serializer("account_update_operation_fee_parameters", { fee: int64,
        price_per_kbyte: uint32 });
 //44
var call_contract_function_operation_fee_parameters = exports.call_contract_function_operation_fee_parameters = new Serializer("call_contract_function_operation_fee_parameters", { fee: int64,
    price_per_kbyte: uint32 });      
    
//46
var register_nh_asset_creator_operation_fee_parameters = exports.register_nh_asset_creator_operation_fee_parameters = new Serializer("account_update_operation_fee_parameters", { fee: int64,
    price_per_kbyte: uint32 });
//47 
var create_world_view_operation_fee_parameters = exports.create_world_view_operation_fee_parameters = new Serializer("create_world_view_operation_fee_parameters", { fee: int64,
    price_per_kbyte: uint32 });

//48
var relate_world_view_operation_fee_parameters = exports.relate_world_view_operation_fee_parameters = new Serializer("relate_world_view_operation_fee_parameters", { fee: int64,
    price_per_kbyte: uint32 });

//49
var create_nh_asset_operation_fee_parameters = exports.create_nh_asset_operation_fee_parameters = new Serializer("create_nh_asset_operation_fee_parameters", { fee: int64,
    price_per_kbyte: uint32 });


var update_collateral_for_gas_operation_fee_parameters = exports.update_collateral_for_gas_operation_fee_parameters = new Serializer("update_collateral_for_gas_operation_fee_parameters", { fee: int64,
    price_per_kbyte: uint32 });

//50
var delete_nh_asset_operation_fee_parameters = exports.delete_nh_asset_operation_fee_parameters = new Serializer("delete_nh_asset_operation_fee_parameters", { fee: int64,
    price_per_kbyte: uint32 });

//51
var transfer_nh_asset_operation_fee_parameters = exports.transfer_nh_asset_operation_fee_parameters = new Serializer("transfer_nh_asset_operation_fee_parameters", { fee: int64,
    price_per_kbyte: uint32 });

//52
var create_nh_asset_order_operation_fee_parameters = exports.create_nh_asset_order_operation_fee_parameters = new Serializer("create_nh_asset_order_operation_fee_parameters", { fee: int64,
    price_per_kbyte: uint32 }); 

//53
var cancel_nh_asset_order_operation_fee_parameters = exports.cancel_nh_asset_order_operation_fee_parameters = new Serializer("cancel_nh_asset_order_operation_fee_parameters", { fee: int64,
    price_per_kbyte: uint32 }); 

//54
var fill_nh_asset_order_operation_fee_parameters = exports.fill_nh_asset_order_operation_fee_parameters = new Serializer("fill_nh_asset_order_operation_fee_parameters", { fee: int64,
    price_per_kbyte: uint32 }); 
//59
var revise_contract_operation_fee_parameters = exports.revise_contract_operation_fee_parameters = new Serializer("revise_contract_operation_fee_parameters", { fee: int64,
    price_per_kbyte: uint32 });    

//60
var crontab_create_operation_fee_parameters = exports.crontab_create_operation_fee_parameters = new Serializer("crontab_create_operation_fee_parameters", { fee: int64,
    price_per_kbyte: uint32 });    

//61
var crontab_cancel_operation_fee_parameters = exports.crontab_cancel_operation_fee_parameters = new Serializer("crontab_cancel_operation_fee_parameters", { fee: int64,
    price_per_kbyte: uint32 }); 
       
//62
var crontab_recover_operation_fee_parameters = exports.crontab_recover_operation_fee_parameters = new Serializer("crontab_recover_operation_fee_parameters", { fee: int64,
    price_per_kbyte: uint32 });    
    

//44
var fba_distribute_operation_fee_parameters = exports.fba_distribute_operation_fee_parameters = new Serializer("fba_distribute_operation_fee_parameters", { fee: int64,
            price_per_kbyte: uint32 });
//45

var bid_collateral_operation_fee_parameters = exports.bid_collateral_operation_fee_parameters = new Serializer("bid_collateral_operation_fee_parameters", { fee: int64,
            price_per_kbyte: uint32 });
            

            
var account_upgrade_operation_fee_parameters = exports.account_upgrade_operation_fee_parameters = new Serializer("account_upgrade_operation_fee_parameters", { membership_annual_fee: uint64,
    membership_lifetime_fee: uint64 });

var account_transfer_operation_fee_parameters = exports.account_transfer_operation_fee_parameters = new Serializer("account_transfer_operation_fee_parameters", { fee: uint64 });

var asset_create_operation_fee_parameters = exports.asset_create_operation_fee_parameters = new Serializer("asset_create_operation_fee_parameters", { symbol3: uint64,
    symbol4: uint64,
    long_symbol: uint64,
    price_per_kbyte: uint32 });

var asset_update_operation_fee_parameters = exports.asset_update_operation_fee_parameters = new Serializer("asset_update_operation_fee_parameters", { fee: uint64,
    price_per_kbyte: uint32 });


    
var asset_update_restricted_operation_fee_parameters = exports.asset_update_restricted_operation_fee_parameters = new Serializer("asset_update_restricted_operation_fee_parameters", { fee: uint64 });

var asset_update_bitasset_operation_fee_parameters = exports.asset_update_bitasset_operation_fee_parameters = new Serializer("asset_update_bitasset_operation_fee_parameters", { fee: uint64 });

var asset_update_feed_producers_operation_fee_parameters = exports.asset_update_feed_producers_operation_fee_parameters = new Serializer("asset_update_feed_producers_operation_fee_parameters", { fee: uint64 });

var asset_issue_operation_fee_parameters = exports.asset_issue_operation_fee_parameters = new Serializer("asset_issue_operation_fee_parameters", { fee: uint64,
    price_per_kbyte: uint32 });

var asset_reserve_operation_fee_parameters = exports.asset_reserve_operation_fee_parameters = new Serializer("asset_reserve_operation_fee_parameters", { fee: uint64 });

var asset_fund_fee_pool_operation_fee_parameters = exports.asset_fund_fee_pool_operation_fee_parameters = new Serializer("asset_fund_fee_pool_operation_fee_parameters", { fee: uint64 });

var asset_settle_operation_fee_parameters = exports.asset_settle_operation_fee_parameters = new Serializer("asset_settle_operation_fee_parameters", { fee: uint64 });

var asset_global_settle_operation_fee_parameters = exports.asset_global_settle_operation_fee_parameters = new Serializer("asset_global_settle_operation_fee_parameters", { fee: uint64 });

var asset_publish_feed_operation_fee_parameters = exports.asset_publish_feed_operation_fee_parameters = new Serializer("asset_publish_feed_operation_fee_parameters", { fee: uint64 });

var witness_create_operation_fee_parameters = exports.witness_create_operation_fee_parameters = new Serializer("witness_create_operation_fee_parameters", { fee: uint64 });

var witness_update_operation_fee_parameters = exports.witness_update_operation_fee_parameters = new Serializer("witness_update_operation_fee_parameters", { fee: int64 });

var proposal_create_operation_fee_parameters = exports.proposal_create_operation_fee_parameters = new Serializer("proposal_create_operation_fee_parameters", { fee: uint64,
    price_per_kbyte: uint32 });

var proposal_update_operation_fee_parameters = exports.proposal_update_operation_fee_parameters = new Serializer("proposal_update_operation_fee_parameters", { fee: uint64,
    price_per_kbyte: uint32 });

var proposal_delete_operation_fee_parameters = exports.proposal_delete_operation_fee_parameters = new Serializer("proposal_delete_operation_fee_parameters", { fee: uint64 });

var withdraw_permission_create_operation_fee_parameters = exports.withdraw_permission_create_operation_fee_parameters = new Serializer("withdraw_permission_create_operation_fee_parameters", { fee: uint64 });

var withdraw_permission_update_operation_fee_parameters = exports.withdraw_permission_update_operation_fee_parameters = new Serializer("withdraw_permission_update_operation_fee_parameters", { fee: uint64 });

var withdraw_permission_claim_operation_fee_parameters = exports.withdraw_permission_claim_operation_fee_parameters = new Serializer("withdraw_permission_claim_operation_fee_parameters", { fee: uint64,
    price_per_kbyte: uint32 });

var withdraw_permission_delete_operation_fee_parameters = exports.withdraw_permission_delete_operation_fee_parameters = new Serializer("withdraw_permission_delete_operation_fee_parameters", { fee: uint64 });

var committee_member_create_operation_fee_parameters = exports.committee_member_create_operation_fee_parameters = new Serializer("committee_member_create_operation_fee_parameters", { fee: uint64 });

var committee_member_update_operation_fee_parameters = exports.committee_member_update_operation_fee_parameters = new Serializer("committee_member_update_operation_fee_parameters", { fee: uint64 });

var committee_member_update_global_parameters_operation_fee_parameters = exports.committee_member_update_global_parameters_operation_fee_parameters = new Serializer("committee_member_update_global_parameters_operation_fee_parameters", { fee: uint64 });

var vesting_balance_create_operation_fee_parameters = exports.vesting_balance_create_operation_fee_parameters = new Serializer("vesting_balance_create_operation_fee_parameters", { fee: uint64 });

var vesting_balance_withdraw_operation_fee_parameters = exports.vesting_balance_withdraw_operation_fee_parameters = new Serializer("vesting_balance_withdraw_operation_fee_parameters", { fee: uint64 });

var worker_create_operation_fee_parameters = exports.worker_create_operation_fee_parameters = new Serializer("worker_create_operation_fee_parameters", { fee: uint64 });

var custom_operation_fee_parameters = exports.custom_operation_fee_parameters = new Serializer("custom_operation_fee_parameters", { fee: uint64,
    price_per_kbyte: uint32 });

var assert_operation_fee_parameters = exports.assert_operation_fee_parameters = new Serializer("assert_operation_fee_parameters", { fee: uint64 });

var balance_claim_operation_fee_parameters = exports.balance_claim_operation_fee_parameters = new Serializer("balance_claim_operation_fee_parameters");

var override_transfer_operation_fee_parameters = exports.override_transfer_operation_fee_parameters = new Serializer("override_transfer_operation_fee_parameters", { fee: uint64,
    price_per_kbyte: uint32 });

var transfer_to_blind_operation_fee_parameters = exports.transfer_to_blind_operation_fee_parameters = new Serializer("transfer_to_blind_operation_fee_parameters", { fee: uint64,
    price_per_output: uint32 });

var blind_transfer_operation_fee_parameters = exports.blind_transfer_operation_fee_parameters = new Serializer("blind_transfer_operation_fee_parameters", { fee: uint64,
    price_per_output: uint32 });

var transfer_from_blind_operation_fee_parameters = exports.transfer_from_blind_operation_fee_parameters = new Serializer("transfer_from_blind_operation_fee_parameters", { fee: uint64 });

var asset_settle_cancel_operation_fee_parameters = exports.asset_settle_cancel_operation_fee_parameters = new Serializer("asset_settle_cancel_operation_fee_parameters");

var asset_claim_fees_operation_fee_parameters = exports.asset_claim_fees_operation_fee_parameters = new Serializer("asset_claim_fees_operation_fee_parameters", { fee: uint64 });

var fee_parameters = static_variant([
    transfer_operation_fee_parameters, limit_order_create_operation_fee_parameters, limit_order_cancel_operation_fee_parameters, call_order_update_operation_fee_parameters, fill_order_operation_fee_parameters, 
    account_create_operation_fee_parameters, account_update_operation_fee_parameters, account_upgrade_operation_fee_parameters, asset_create_operation_fee_parameters, asset_update_operation_fee_parameters,

    asset_update_restricted_operation_fee_parameters,asset_update_bitasset_operation_fee_parameters,asset_update_feed_producers_operation_fee_parameters, asset_issue_operation_fee_parameters, asset_reserve_operation_fee_parameters,
    // asset_fund_fee_pool_operation_fee_parameters,
    asset_settle_operation_fee_parameters, asset_global_settle_operation_fee_parameters,
     asset_publish_feed_operation_fee_parameters, asset_publish_feed_operation_fee_parameters,//witness_create_operation_fee_parameters,
     witness_update_operation_fee_parameters,
   proposal_create_operation_fee_parameters, proposal_update_operation_fee_parameters, proposal_delete_operation_fee_parameters, withdraw_permission_create_operation_fee_parameters, withdraw_permission_update_operation_fee_parameters, 
   withdraw_permission_claim_operation_fee_parameters, withdraw_permission_delete_operation_fee_parameters, committee_member_create_operation_fee_parameters, committee_member_update_operation_fee_parameters, committee_member_update_global_parameters_operation_fee_parameters, vesting_balance_create_operation_fee_parameters, vesting_balance_withdraw_operation_fee_parameters, worker_create_operation_fee_parameters, balance_claim_operation_fee_parameters, transfer_to_blind_operation_fee_parameters, blind_transfer_operation_fee_parameters, transfer_from_blind_operation_fee_parameters, asset_settle_cancel_operation_fee_parameters,
    asset_claim_fees_operation_fee_parameters,fba_distribute_operation_fee_parameters, bid_collateral_operation_fee_parameters,
    contract_create_operation_fee_parameters,call_contract_function_operation_fee_parameters,call_contract_function_operation_fee_parameters,
    register_nh_asset_creator_operation_fee_parameters,create_world_view_operation_fee_parameters,relate_world_view_operation_fee_parameters,create_nh_asset_operation_fee_parameters,delete_nh_asset_operation_fee_parameters,
    transfer_nh_asset_operation_fee_parameters,create_nh_asset_order_operation_fee_parameters,cancel_nh_asset_order_operation_fee_parameters,fill_nh_asset_order_operation_fee_parameters,revise_contract_operation_fee_parameters,
    crontab_create_operation_fee_parameters,crontab_cancel_operation_fee_parameters,crontab_recover_operation_fee_parameters]);
var fee_schedule = exports.fee_schedule = new Serializer("fee_schedule", { parameters: set(fee_parameters),
    scale: uint32 });

var void_result = exports.void_result = new Serializer("void_result");

var asset = exports.asset = new Serializer("asset", { amount: int64,
    asset_id: protocol_id_type("asset") });

var operation_result = static_variant([void_result, object_id_type, asset]);
var processed_transaction = exports.processed_transaction = new Serializer("processed_transaction", { ref_block_num: uint16,
    ref_block_prefix: uint32,
    expiration: time_point_sec,
    operations: array(operation),
    extensions: set(future_extensions),
    signatures: array(bytes(65)),
    operation_results: array(operation_result) });

var signed_block = exports.signed_block = new Serializer("signed_block", { previous: bytes(20),
    timestamp: time_point_sec,
    witness: protocol_id_type("witness"),
    transaction_merkle_root: bytes(20),
    extensions: set(future_extensions),
    witness_signature: bytes(65),
    transactions: array(processed_transaction) });

var block_header = exports.block_header = new Serializer("block_header", { previous: bytes(20),
    timestamp: time_point_sec,
    witness: protocol_id_type("witness"),
    transaction_merkle_root: bytes(20),
    extensions: set(future_extensions) });

var signed_block_header = exports.signed_block_header = new Serializer("signed_block_header", { previous: bytes(20),
    timestamp: time_point_sec,
    witness: protocol_id_type("witness"),
    transaction_merkle_root: bytes(20),
    extensions: set(future_extensions),
    witness_signature: bytes(65) });

var memo_data = exports.memo_data = new Serializer("memo_data", { from: public_key,
    to: public_key,
    nonce: uint64,
    message: bytes() });

var memo_type = static_variant([string, memo_data]);

var transfer = exports.transfer = new Serializer("transfer", { 
    from: protocol_id_type("account"),
    to: protocol_id_type("account"),
    amount: asset,
    memo: optional(memo_type),
    extensions: set(future_extensions) });
      
var limit_order_create = exports.limit_order_create = new Serializer("limit_order_create", {
    seller: protocol_id_type("account"),
    amount_to_sell: asset,
    min_to_receive: asset,
    expiration: time_point_sec,
    fill_or_kill: bool,
    extensions: set(future_extensions) });

var limit_order_cancel = exports.limit_order_cancel = new Serializer("limit_order_cancel", {
    fee_paying_account: protocol_id_type("account"),
    order: protocol_id_type("limit_order"),
    extensions: set(future_extensions) });

var call_order_update = exports.call_order_update = new Serializer("call_order_update", {
     //fee: asset,
    funding_account: protocol_id_type("account"),
    delta_collateral: asset,
    delta_debt: asset,
    extensions: set(future_extensions) });

var fill_order = exports.fill_order = new Serializer("fill_order", {
    order_id: object_id_type,
    account_id: protocol_id_type("account"),
    pays: asset,
    receives: asset });

var authority = exports.authority = new Serializer("authority", { weight_threshold: uint32,
    account_auths: map(protocol_id_type("account"), uint16),
    key_auths: map(public_key, uint16),
    address_auths: map(address, uint16) });

var account_options = exports.account_options = new Serializer("account_options", { memo_key: public_key,
    // voting_account: protocol_id_type("account"),
    // num_witness: uint16,
    // num_committee: uint16,
    votes: set(vote_id),
    extensions: set(future_extensions) });

var account_create = exports.account_create = new Serializer("account_create", {
    registrar: protocol_id_type("account"),
    // referrer: protocol_id_type("account"),
    // referrer_percent: uint16,
    name: string,
    owner: authority,
    active: authority,
    options: account_options,
    extensions: set(future_extensions) });

var account_update = exports.account_update = new Serializer("account_update", {
    lock_with_vote:optional(pair(uint32,asset)),
    account: protocol_id_type("account"),
    owner: optional(authority),
    active: optional(authority),
    new_options: optional(account_options),
    extensions: set(future_extensions) });

 //45
var execute_bid = exports.execute_bid = new Serializer("execute_bid", { 
    owner: protocol_id_type("account"),
    name: string,
    data: string,
    extensions: set(future_extensions) });

//46
var contract_create = exports.contract_create = new Serializer("contract_create", { 
    owner: protocol_id_type("account"),
    name: string,
    data: string,
    contract_authority:public_key,
    extensions: set(future_extensions) });

// var vector_lua_types = lua_types([int64, double, string, bool,table(lua_types([int64,double,string,bool])),optional(memo_data),asset]);

var call_contract_function = exports.call_contract_function = new Serializer("call_contract_function", {
    caller: protocol_id_type("account"),
    contract_id: protocol_id_type("contract_id_type"),
    function_name: string,
    value_list:array(lua_types),
    extensions: set(future_extensions) });

var register_nh_asset_creator = exports.register_nh_asset_creator = new Serializer("register_nh_asset_creator", {
    fee_paying_account: protocol_id_type("account")});

var create_world_view = exports.create_world_view = new Serializer("create_world_view", {
    fee_paying_account: protocol_id_type("account"),
    world_view:string
});

var relate_world_view = exports.relate_world_view = new Serializer("relate_world_view", {
    related_account: protocol_id_type("account"),
    world_view:string,
    view_owner:protocol_id_type("account")
});

// var _base_describe = static_variant([void_result, object_id_type, asset]);

var create_nh_asset = exports.create_nh_asset = new Serializer("create_nh_asset", {
    fee_paying_account: protocol_id_type("account"),
    owner:protocol_id_type("account"),
    asset_id:string,
    world_view:string,
    base_describe:string
});

var update_collateral_for_gas = exports.update_collateral_for_gas = new Serializer("update_collateral_for_gas", {
    mortgager: protocol_id_type("account"),
    beneficiary:protocol_id_type("account"),
    collateral:int64
});

var delete_nh_asset = exports.delete_nh_asset = new Serializer("delete_nh_asset", {
    fee_paying_account: protocol_id_type("account"),
    nh_asset:nh_protocol_id_type("nh_item_object_type")
});

var transfer_nh_asset = exports.transfer_nh_asset = new Serializer("transfer_nh_asset", { 
    from: protocol_id_type("account"),
    to:protocol_id_type("account"),
    nh_asset:nh_protocol_id_type("nh_item_object_type")
});

var create_nh_asset_order = exports.create_nh_asset_order = new Serializer("create_nh_asset_order", {
    seller: protocol_id_type("account"),
    otcaccount:protocol_id_type("account"),
    pending_orders_fee:asset,
    nh_asset:nh_protocol_id_type("nh_item_object_type"),
    memo:string,
    price:asset,
    expiration:time_point_sec
});


var cancel_nh_asset_order=exports.cancel_nh_asset_order=new Serializer("cancel_nh_asset_order", {
    order: nh_protocol_id_type("nh_item_order_object_type"),
    fee_paying_account:protocol_id_type("account"),
    extensions: set(future_extensions) 
});


var fill_nh_asset_order=exports.fill_nh_asset_order=new Serializer("fill_nh_asset_order", {
    order: nh_protocol_id_type("nh_item_order_object_type"),
    fee_paying_account:protocol_id_type("account"),
    seller:protocol_id_type("account"),
    nh_asset:nh_protocol_id_type("nh_item_object_type"),
    price_amount:string,
    price_asset_id:protocol_id_type("asset"),
    price_asset_symbol:string,
    extensions: set(future_extensions) 
});


var revise_contract=exports.revise_contract=new Serializer("revise_contract", {
    reviser:protocol_id_type("account"),
    contract_id: protocol_id_type("contract_id_type"),
    data:string,
    extensions: set(future_extensions) 
});

var op_wrapper = exports.op_wrapper = new Serializer("op_wrapper", { op: operation });


var crontab_create=exports.crontab_create=new Serializer("crontab_create", { fee: asset,
    crontab_creator:protocol_id_type("account"),
    crontab_ops: array(op_wrapper),
    start_time:time_point_sec,
    execute_interval:uint64,
    scheduled_execute_times:uint64,
    extensions: set(future_extensions) 
});

var crontab_cancel=exports.crontab_cancel=new Serializer("crontab_cancel", { fee: asset,
    fee_paying_account:protocol_id_type("account"),
    task: protocol_id_type("crontab_id_type"),
    extensions: set(future_extensions) 
});


var crontab_recover=exports.crontab_recover=new Serializer("crontab_recover", { fee: asset,
    crontab_owner:protocol_id_type("account"),
    crontab: protocol_id_type("crontab_id_type"),
    restart_time:time_point_sec,
    extensions: set(future_extensions) 
});

var fba_distribute = exports.fba_distribute = new Serializer("fba_distribute", {
    owner: protocol_id_type("account"),
    name: string,
    data: string,
    extensions: set(future_extensions) });

var bid_collateral = exports.bid_collateral = new Serializer("bid_collateral", {
    owner: protocol_id_type("account"),
    name: string,
    data: string,
    extensions: set(future_extensions) });


var account_upgrade = exports.account_upgrade = new Serializer("account_upgrade", { 
    account_to_upgrade: protocol_id_type("account"),
    upgrade_to_lifetime_member: bool,
    extensions: set(future_extensions) });

var account_transfer = exports.account_transfer = new Serializer("account_transfer", { 
    account_id: protocol_id_type("account"),
    new_owner: protocol_id_type("account"),
    extensions: set(future_extensions) });

var price = exports.price = new Serializer("price", { base: asset,
    quote: asset });

var asset_options = exports.asset_options = new Serializer("asset_options", { max_supply: int64,
    market_fee_percent: uint16,
    max_market_fee: int64,
    issuer_permissions: uint16,
    flags: uint16,
    core_exchange_rate: optional(price),//optional(price)
    description: string,
    extensions: set(future_extensions) });

var bitasset_options = exports.bitasset_options = new Serializer("bitasset_options", { feed_lifetime_sec: uint32,
    minimum_feeds: uint8,
    force_settlement_delay_sec: uint32,
    force_settlement_offset_percent: uint16,
    maximum_force_settlement_volume: uint16,
    short_backing_asset: protocol_id_type("asset"),
    extensions: set(future_extensions) });

var asset_create = exports.asset_create = new Serializer("asset_create", {
    issuer: protocol_id_type("account"),
    symbol: string,
    precision: uint8,
    common_options: asset_options,
    bitasset_opts: optional(bitasset_options),
    // is_prediction_market: bool,
    extensions: set(future_extensions) });

var asset_update = exports.asset_update = new Serializer("asset_update", { //fee: asset,
    issuer: protocol_id_type("account"),
    asset_to_update: protocol_id_type("asset"),
    new_issuer: optional(protocol_id_type("account")),
    new_options: asset_options,
    extensions: set(future_extensions) });

      
var asset_update_restricted = exports.asset_update_restricted = new Serializer("asset_update_restricted", {
    payer: protocol_id_type("account"),
    target_asset: protocol_id_type("asset"),
    isadd: bool,
    restricted_type:uint8,
    restricted_list:array(object_id_type),
    extensions: set(future_extensions) });   

var asset_update_bitasset = exports.asset_update_bitasset = new Serializer("asset_update_bitasset", { //fee: asset,
    issuer: protocol_id_type("account"),
    asset_to_update: protocol_id_type("asset"),
    new_options: bitasset_options,
    extensions: set(future_extensions) });

var asset_update_feed_producers = exports.asset_update_feed_producers = new Serializer("asset_update_feed_producers", {
    issuer: protocol_id_type("account"),
    asset_to_update: protocol_id_type("asset"),
    new_feed_producers: set(protocol_id_type("account")),
    extensions: set(future_extensions) });

var asset_issue = exports.asset_issue = new Serializer("asset_issue", { 
    issuer: protocol_id_type("account"),
    asset_to_issue: asset,
    issue_to_account: protocol_id_type("account"),
    memo: optional(memo_type),
    extensions: set(future_extensions) });

var asset_reserve = exports.asset_reserve = new Serializer("asset_reserve", {
    payer: protocol_id_type("account"),
    amount_to_reserve: asset,
    extensions: set(future_extensions) });

var asset_fund_fee_pool = exports.asset_fund_fee_pool = new Serializer("asset_fund_fee_pool", {
    from_account: protocol_id_type("account"),
    asset_id: protocol_id_type("asset"),
    amount: int64,
    extensions: set(future_extensions) });

var asset_settle = exports.asset_settle = new Serializer("asset_settle", { 
    account: protocol_id_type("account"),
    amount: asset,
    extensions: set(future_extensions) });

var asset_global_settle = exports.asset_global_settle = new Serializer("asset_global_settle", { 
    issuer: protocol_id_type("account"),
    asset_to_settle: protocol_id_type("asset"),
    settle_price: price,
    extensions: set(future_extensions) });

var price_feed = exports.price_feed = new Serializer("price_feed", { settlement_price: price,
    maintenance_collateral_ratio: uint16,
    maximum_short_squeeze_ratio: uint16,
    //core_exchange_rate: price//
   });

var asset_publish_feed = exports.asset_publish_feed = new Serializer("asset_publish_feed", { //fee: asset,//
    publisher: protocol_id_type("account"),
    asset_id: protocol_id_type("asset"),
    feed: price_feed,
    extensions: set(future_extensions) });

var witness_create = exports.witness_create = new Serializer("witness_create", { //fee: asset,
    witness_account: protocol_id_type("account"),
    url: string,
    block_signing_key: public_key });

var witness_update = exports.witness_update = new Serializer("witness_update", {
    witness: protocol_id_type("witness"),
    witness_account: protocol_id_type("account"),
    new_url: optional(string),
    new_signing_key: optional(public_key),
    work_status:bool });


var proposal_create = exports.proposal_create = new Serializer("proposal_create", { 
    fee_paying_account: protocol_id_type("account"),
    expiration_time: time_point_sec,
    proposed_ops: array(op_wrapper),
    review_period_seconds: optional(uint32),
    extensions: set(future_extensions) });

var proposal_update = exports.proposal_update = new Serializer("proposal_update", { 
    fee_paying_account: protocol_id_type("account"),
    proposal: protocol_id_type("proposal"),
    active_approvals_to_add: set(protocol_id_type("account")),
    active_approvals_to_remove: set(protocol_id_type("account")),
    owner_approvals_to_add: set(protocol_id_type("account")),
    owner_approvals_to_remove: set(protocol_id_type("account")),
    key_approvals_to_add: set(public_key),
    key_approvals_to_remove: set(public_key),
    extensions: set(future_extensions) });

var proposal_delete = exports.proposal_delete = new Serializer("proposal_delete", {
    fee_paying_account: protocol_id_type("account"),
    using_owner_authority: bool,
    proposal: protocol_id_type("proposal"),
    extensions: set(future_extensions) });

var withdraw_permission_create = exports.withdraw_permission_create = new Serializer("withdraw_permission_create", { fee: asset,
    withdraw_from_account: protocol_id_type("account"),
    authorized_account: protocol_id_type("account"),
    withdrawal_limit: asset,
    withdrawal_period_sec: uint32,
    periods_until_expiration: uint32,
    period_start_time: time_point_sec });

var withdraw_permission_update = exports.withdraw_permission_update = new Serializer("withdraw_permission_update", { fee: asset,
    withdraw_from_account: protocol_id_type("account"),
    authorized_account: protocol_id_type("account"),
    permission_to_update: protocol_id_type("withdraw_permission"),
    withdrawal_limit: asset,
    withdrawal_period_sec: uint32,
    period_start_time: time_point_sec,
    periods_until_expiration: uint32 });

var withdraw_permission_claim = exports.withdraw_permission_claim = new Serializer("withdraw_permission_claim", { fee: asset,
    withdraw_permission: protocol_id_type("withdraw_permission"),
    withdraw_from_account: protocol_id_type("account"),
    withdraw_to_account: protocol_id_type("account"),
    amount_to_withdraw: asset,
    memo: optional(memo_data) });

var withdraw_permission_delete = exports.withdraw_permission_delete = new Serializer("withdraw_permission_delete", { fee: asset,
    withdraw_from_account: protocol_id_type("account"),
    authorized_account: protocol_id_type("account"),
    withdrawal_permission: protocol_id_type("withdraw_permission") });

var committee_member_create = exports.committee_member_create = new Serializer("committee_member_create", {
    committee_member_account: protocol_id_type("account"),
    url: string });

var committee_member_update = exports.committee_member_update = new Serializer("committee_member_update", {
    committee_member: protocol_id_type("committee_member"),
    committee_member_account: protocol_id_type("account"),
    new_url: optional(string),
    work_status:bool });

var chain_parameters = exports.chain_parameters = new Serializer("chain_parameters", { current_fees: fee_schedule,
    block_interval: uint8,
    maintenance_interval: uint32,
    maintenance_skip_slots: uint8,
    committee_proposal_review_period: uint32,
    maximum_transaction_size: uint32,
    maximum_block_size: uint32,
    maximum_time_until_expiration: uint32,
    maximum_proposal_lifetime: uint32,
    maximum_asset_whitelist_authorities: uint8,
    maximum_asset_feed_publishers: uint8,
    maximum_witness_count: uint16,
    maximum_committee_count: uint16,
    maximum_authority_membership: uint16,
    reserve_percent_of_fee: uint16,
    network_percent_of_fee: uint16,
    lifetime_referrer_percent_of_fee: uint16,
    cashback_vesting_period_seconds: uint32,
    cashback_vesting_threshold: int64,
    count_non_member_votes: bool,
    allow_non_member_whitelists: bool,
    witness_pay_per_block: int64,
    worker_budget_per_day: int64,
    max_predicate_opcode: uint16,
    fee_liquidation_threshold: int64,
    accounts_per_fee_scale: uint16,
    account_fee_scale_bitshifts: uint8,
    max_authority_depth: uint8,
    extensions: set(future_extensions) });

var committee_member_update_global_parameters = exports.committee_member_update_global_parameters = new Serializer("committee_member_update_global_parameters", { fee: asset,
    new_parameters: chain_parameters });

var linear_vesting_policy_initializer = exports.linear_vesting_policy_initializer = new Serializer("linear_vesting_policy_initializer", { begin_timestamp: time_point_sec,
    vesting_cliff_seconds: uint32,
    vesting_duration_seconds: uint32 });

var cdd_vesting_policy_initializer = exports.cdd_vesting_policy_initializer = new Serializer("cdd_vesting_policy_initializer", { start_claim: time_point_sec,
    vesting_seconds: uint32 });

var vesting_policy_initializer = static_variant([linear_vesting_policy_initializer, cdd_vesting_policy_initializer]);

var vesting_balance_create = exports.vesting_balance_create = new Serializer("vesting_balance_create", { fee: asset,
    creator: protocol_id_type("account"),
    owner: protocol_id_type("account"),
    amount: asset,
    policy: vesting_policy_initializer });

var vesting_balance_withdraw = exports.vesting_balance_withdraw = new Serializer("vesting_balance_withdraw", {
    vesting_balance: protocol_id_type("vesting_balance"),
    owner: protocol_id_type("account"),
    amount: asset });

var refund_worker_initializer = exports.refund_worker_initializer = new Serializer("refund_worker_initializer");

var vesting_balance_worker_initializer = exports.vesting_balance_worker_initializer = new Serializer("vesting_balance_worker_initializer", { pay_vesting_period_days: uint16 });

var burn_worker_initializer = exports.burn_worker_initializer = new Serializer("burn_worker_initializer");

var worker_initializer = static_variant([refund_worker_initializer, vesting_balance_worker_initializer, burn_worker_initializer]);

var worker_create = exports.worker_create = new Serializer("worker_create", { fee: asset,
    owner: protocol_id_type("account"),
    work_begin_date: time_point_sec,
    work_end_date: time_point_sec,
    daily_pay: int64,
    name: string,
    url: string,
    initializer: worker_initializer });

var custom = exports.custom = new Serializer("custom", { fee: asset,
    payer: protocol_id_type("account"),
    required_auths: set(protocol_id_type("account")),
    id: uint16,
    data: bytes() });

var account_name_eq_lit_predicate = exports.account_name_eq_lit_predicate = new Serializer("account_name_eq_lit_predicate", { account_id: protocol_id_type("account"),
    name: string });

var asset_symbol_eq_lit_predicate = exports.asset_symbol_eq_lit_predicate = new Serializer("asset_symbol_eq_lit_predicate", { asset_id: protocol_id_type("asset"),
    symbol: string });

var block_id_predicate = exports.block_id_predicate = new Serializer("block_id_predicate", { id: bytes(20) });

var predicate = static_variant([account_name_eq_lit_predicate, asset_symbol_eq_lit_predicate, block_id_predicate]);

var assert = exports.assert = new Serializer("assert", { fee: asset,
    fee_paying_account: protocol_id_type("account"),
    predicates: array(predicate),
    required_auths: set(protocol_id_type("account")),
    extensions: set(future_extensions) });

var balance_claim = exports.balance_claim = new Serializer("balance_claim", { fee: asset,
    deposit_to_account: protocol_id_type("account"),
    balance_to_claim: protocol_id_type("balance"),
    balance_owner_key: public_key,
    total_claimed: asset });

var override_transfer = exports.override_transfer = new Serializer("override_transfer", { fee: asset,
    issuer: protocol_id_type("account"),
    from: protocol_id_type("account"),
    to: protocol_id_type("account"),
    amount: asset,
    memo: optional(memo_data),
    extensions: set(future_extensions) });

var stealth_confirmation = exports.stealth_confirmation = new Serializer("stealth_confirmation", { one_time_key: public_key,
    to: optional(public_key),
    encrypted_memo: bytes() });

var blind_output = exports.blind_output = new Serializer("blind_output", { commitment: bytes(33),
    range_proof: bytes(),
    owner: authority,
    stealth_memo: optional(stealth_confirmation) });

var transfer_to_blind = exports.transfer_to_blind = new Serializer("transfer_to_blind", { fee: asset,
    amount: asset,
    from: protocol_id_type("account"),
    blinding_factor: bytes(32),
    outputs: array(blind_output) });

var blind_input = exports.blind_input = new Serializer("blind_input", { commitment: bytes(33),
    owner: authority });

var blind_transfer = exports.blind_transfer = new Serializer("blind_transfer", {
    inputs: array(blind_input),
    outputs: array(blind_output) });

var transfer_from_blind = exports.transfer_from_blind = new Serializer("transfer_from_blind", { 
    amount: asset,
    to: protocol_id_type("account"),
    blinding_factor: bytes(32),
    inputs: array(blind_input) });

var asset_settle_cancel = exports.asset_settle_cancel = new Serializer("asset_settle_cancel", { fee: asset,
    settlement: protocol_id_type("force_settlement"),
    account: protocol_id_type("account"),
    amount: asset,
    extensions: set(future_extensions) });

var asset_claim_fees = exports.asset_claim_fees = new Serializer("asset_claim_fees", {
    issuer: protocol_id_type("account"),
    amount_to_claim: asset,
    extensions: set(future_extensions) });

operation.st_operations = [
    transfer, limit_order_create, limit_order_cancel, call_order_update, fill_order, 
    account_create, account_update, account_upgrade, asset_create, asset_update,
    asset_update_restricted,asset_update_bitasset,asset_update_feed_producers, asset_issue, asset_reserve,
    asset_settle, asset_global_settle, asset_publish_feed,witness_create, witness_update,
    proposal_create,proposal_update,proposal_delete,committee_member_create,committee_member_update,
    // witness_update, 
    vesting_balance_create,committee_member_update_global_parameters,vesting_balance_withdraw,worker_create, balance_claim, 
    asset_settle_cancel, asset_claim_fees,bid_collateral,execute_bid,contract_create,
    call_contract_function,call_contract_function, register_nh_asset_creator,create_world_view,relate_world_view,
    create_nh_asset,delete_nh_asset,transfer_nh_asset,create_nh_asset_order,cancel_nh_asset_order,
    fill_nh_asset_order,46,47,48,49,
    revise_contract,crontab_create,crontab_cancel,crontab_recover,
    update_collateral_for_gas];

var transaction = exports.transaction = new Serializer("transaction", { ref_block_num: uint16,
    ref_block_prefix: uint32,
    expiration: time_point_sec,
    operations: array(operation),
    extensions: set(future_extensions) });

var signed_transaction = exports.signed_transaction = new Serializer("signed_transaction", { ref_block_num: uint16,
    ref_block_prefix: uint32,
    expiration: time_point_sec,
    operations: array(operation),
    extensions: set(future_extensions),
    signatures: array(bytes(65)) });
//# -------------------------------
//#  Generated code end
//# -------------------------------

// Custom Types

var stealth_memo_data = exports.stealth_memo_data = new Serializer("stealth_memo_data", {
    from: optional(public_key),
    amount: asset,
    blinding_factor: bytes(32),
    commitment: bytes(33),
    check: uint32
});
// var stealth_confirmation = new Serializer(
//     "stealth_confirmation", {
//     one_time_key: public_key,
//     to: optional( public_key ),
//     encrypted_memo: stealth_memo_data
// })
