// bcxjs-ws/cjs/src/ChainConfig.js
"use strict";

exports.__esModule = true;
var _this = void 0;

var ecc_config = {
    address_prefix: process.env.npm_config__graphene_ecc_default_address_prefix || "COCOS"
};

_this = {
    chain_id:"",
    core_asset: "COCOS",
    address_prefix: "COCOS",
    expire_in_secs: 15,
    expire_in_secs_proposal: 24 * 60 * 60,
    review_in_secs_committee: 24 * 60 * 60,
    networks:[],

    /** Set a few properties for known chain IDs. */
    setChainId: function setChainId(chain_id) {

        var i = void 0,
            len = void 0,
            network = void 0,    
            network_name = void 0,
            ref = void 0;
        for (i = 0, len = _this.networks.length; i < len; i++) {
            network=_this.networks[i];
            network_name = network.core_asset;
            if (network.chain_id === chain_id) {
                _this.network_name = network_name;
                if (network.core_asset) {
                    _this.core_asset = network.core_asset;
                    _this.address_prefix = network.core_asset;
                    ecc_config.address_prefix = network.core_asset;
                }
                _this.chain_id=chain_id;
                // console.log("INFO    Configured for", network_name, ":", network.core_asset, "\n");
                return {
                    network_name: network_name,
                    network: network
                };
            }
        }
        if (!_this.network_name) {
            console.log("Unknown chain id (this may be a testnet)", chain_id);
        }
    },

    reset: function reset() {
        _this.core_asset = "COCOS";
        _this.address_prefix = "COCOS";
        ecc_config.address_prefix = "COCOS";
        _this.expire_in_secs = 15;
        _this.expire_in_secs_proposal = 24 * 60 * 60;

        console.log("Chain config reset");
    },

    setPrefix: function setPrefix() {
        var prefix = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "COCOS";

        _this.address_prefix = prefix;
        ecc_config.address_prefix = prefix;
    }
};

exports.default = _this;
module.exports = exports["default"];
