// bcxjs-cores/dist/chain/src/ChainTypes.js
"use strict";

exports.__esModule = true;
var ChainTypes = {};

ChainTypes.reserved_spaces = {
    relative_protocol_ids: 0,
    protocol_ids: 1,
    implementation_ids: 2,
    nh_protocol_ids:4
};

ChainTypes.object_type = {
    "null": 0,
    base: 1,
    account: 2,
    asset: 3,
    force_settlement: 4,
    committee_member: 5,
    witness: 6,
    limit_order: 7,
    call_order: 8,
    custom: 9,
    proposal: 10,
    operation_history: 11,
    withdraw_permission: 12,
    vesting_balance: 13,
    worker: 14,
    balance: 15,
    contract_id_type: 16,
    crontab_id_type:19
};

ChainTypes.lua_type = {
    int:0,
    number:1,
    string:2,
    boolean:3,
    table:4,
    function:5,
    memo_data:6,
    asset:7
};

ChainTypes.impl_object_type = {
    global_property: 0,
    dynamic_global_property: 1,
    index_meta: 2,
    asset_dynamic_data: 3,
    asset_bitasset_data: 4,
    account_balance: 5,
    account_statistics: 6,
    transaction: 7,
    block_summary: 8,
    account_transaction_history: 9,
    blinded_balance: 10,
    chain_property: 11,
    witness_schedule: 12,
    budget_record: 13
};

ChainTypes.nh_object_type = {
    nh_developer_object_type: 0,
    nh_version_object_type:1,
    nh_item_object_type:2,
    nh_item_order_object_type:3
};

ChainTypes.vote_type = {
    committee: 0,
    witness: 1,
    worker: 2
};

ChainTypes.operations = {
    transfer: 0,
    limit_order_create: 1,
    limit_order_cancel: 2,
    call_order_update: 3,
    fill_order: 4,
    account_create: 5,
    account_update: 6,
    account_upgrade: 7,
    asset_create: 8,
    asset_update: 9,
    asset_update_restricted:10,
    asset_update_bitasset: 11,
    asset_update_feed_producers: 12,
    asset_issue: 13,
    asset_reserve: 14,
    asset_settle: 15,
    asset_global_settle: 16,
    asset_publish_feed: 17,
    witness_create: 18,
    // asset_publish_feed: 18,

    witness_update: 19,
    proposal_create:20,
    proposal_update: 21,
    proposal_delete: 22,
    committee_member_create: 23,
    committee_member_update: 24,
    committee_member_update_global_parameters: 25,
    vesting_balance_create: 26,
    vesting_balance_withdraw: 27,
    worker_create: 28,
    balance_claim: 29,
    asset_settle_cancel: 30,
    asset_claim_fees: 31,
    bid_collateral:32,
    execute_bid:33,
    contract_create:34,
    call_contract_function:35,
    temporary_authority_chang:36,
    register_nh_asset_creator:37,
    create_world_view:38,
    relate_world_view:39,
    create_nh_asset:40,
    delete_nh_asset:41,
    transfer_nh_asset:42,
    create_nh_asset_order:43,
    cancel_nh_asset_order:44,
    fill_nh_asset_order:45,
    create_file:46,
    add_file_relate_account:47,
    file_signature:48,
    relate_parent_file:49,
    revise_contract:50,
    crontab_create:51, 
    crontab_cancel:52,
    crontab_recover:53,
    update_collateral_for_gas:54    
};

exports.default = ChainTypes;
module.exports = exports["default"];
